### ISSUE TITLE
[dtest] compaction_additional_test.TestTimeWindowDataSegregation.test_rebuild_node_streaming fails 

### ISSUE THREAD

**Issue by asias** [scylladb/scylladb#18809]
Found in CI: https://github.com/scylladb/scylladb/pull/18780#issuecomment-2121861094

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1652/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/FullDtest___full_split005___test_rebuild_node_streaming/

```
self = <compaction_additional_test.TestTimeWindowDataSegregation object at 0x7f87648aa090>

    @pytest.mark.required_features("!tablets")
    def test_rebuild_node_streaming(self):
        [node1, _, node3], session = self.prepare(3)
        self._create_ks_cl_with_twcs(session, rf=3)
    
        self._simulate_write_process_in_minutes(session, duration_minutes=10, flush_period_seconds=20)
        for node in self.cluster.nodelist():
            self._check_sstable_timestamps(node)
        # stop node and remove all data
        node3 = self.cluster.nodelist()[2]  # type: ScyllaNode
        node3.stop()
        data_dir = os.path.join(node3.get_path(), "data", self.keyspace_name)
        shutil.rmtree(data_dir, ignore_errors=True)
        # start node and rebuild
        node3.start(wait_other_notice=True, wait_for_binary_proto=True)
        mark = node3.mark_log()
        node3.nodetool("rebuild")
        node3.watch_log_for("Streaming for rebuild successful|" "rebuild_with_repair: finished with keyspace=ks", from_mark=mark)
        run_rest_api(node3, f"/storage_service/keyspace_offstrategy_compaction/{self.keyspace_name}", params={"cf": self.table_name})
        node3.watch_log_for(f"Done with off-strategy compaction for {self.keyspace_name}", from_mark=mark)
>       self._check_sstable_timestamps(node3)

compaction_additional_test.py:1850: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compaction_additional_test.py:1400: in _check_sstable_timestamps
    stats = self.get_stats(node, sf, keyspace_name=self.keyspace_name, table_name=self.table_name)
compaction_additional_test.py:81: in get_stats
    metadata = node.dump_sstable_stats(keyspace=keyspace_name, column_family=table_name, datafiles=[statistics_file])
../scylla/.local/lib/python3.11/site-packages/ccmlib/scylla_node.py:1666: in dump_sstable_stats
    sstable_stats = self.run_scylla_sstable('dump-statistics',
../scylla/.local/lib/python3.11/site-packages/ccmlib/scylla_node.py:1517: in run_scylla_sstable
    ret[""] = do_invoke(sstables)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

sstables = ['/jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.dtest/dtest-ntndu5y1/test/node3/data/ks/test-6a95bce0172d11efa5f6fdcbcce57e18/md-3ggb_0df2_3qlog2o8h8if1c9ts2-big-Statistics.db']

    def do_invoke(sstables):
        # there are chances that the table is not replicated on this node,
        # in that case, scylla tool will fail to dump the sstables and error
        # out. let's just return an empty list for the partitions, so that
        # dump_sstables() can still parse it in the same way.
        if not sstables:
            empty_dump = {'sstables': {'anonymous': []}}
            stdout, stderr = json.dumps(empty_dump), ''
            if text:
                return stdout, stderr
            else:
                return stdout.encode('utf-8'), stderr.encode('utf-8')
        common_args = [scylla_path, "sstable", command] + additional_args
        env = self._get_environ()
        res = subprocess.run(common_args + sstables, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=text, check=False, env=env)
        if res.returncode:
>           raise ToolError(command=' '.join(common_args + sstables), exit_status=res.returncode, stdout=res.stdout, stderr=res.stderr)
E           ccmlib.node.ToolError: Subprocess /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.dtest/dtest-ntndu5y1/test/node3/bin/scylla sstable dump-statistics /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.dtest/dtest-ntndu5y1/test/node3/data/ks/test-6a95bce0172d11efa5f6fdcbcce57e18/md-3ggb_0df2_3qlog2o8h8if1c9ts2-big-Statistics.db exited with non-zero status; exit status: 1; 
E           stderr: ERROR 2024-05-21 04:49:51,169 [shard 0:main] sstable - Could not create SSTable component /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.dtest/dtest-ntndu5y1/test/node3/data/ks/test-6a95bce0172d11efa5f6fdcbcce57e18/../../system_schema/views-9786ac1cdd583201a7cdad556410c985/md-3ggb_0df3_04abk1zo4u8i0rn75u-big-TOC.txt. Found exception: std::filesystem::__cxx11::filesystem_error (error system:2, filesystem error: open failed: No such file or directory [/jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.dtest/dtest-ntndu5y1/test/node3/data/ks/test-6a95bce0172d11efa5f6fdcbcce57e18/../../system_schema/views-9786ac1cdd583201a7cdad556410c985/md-3ggb_0df3_04abk1zo4u8i0rn75u-big-TOC.txt])
E           ERROR 2024-05-21 04:49:51,169 [shard 0:main] sstable - Could not create SSTable component /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.dtest/dtest-ntndu5y1/test/node3/data/ks/test-6a95bce0172d11efa5f6fdcbcce57e18/../../system_schema/views-9786ac1cdd583201a7cdad556410c985/md-3ggb_0ddf_4mbnk2gcdignlgbr2a-big-TOC.txt. Found exception: std::filesystem::__cxx11::filesystem_error (error system:2, filesystem error: open failed: No such file or directory [/jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.dtest/dtest-ntndu5y1/test/node3/data/ks/test-6a95bce0172d11efa5f6fdcbcce57e18/../../system_schema/views-9786ac1cdd583201a7cdad556410c985/md-3ggb_0ddf_4mbnk2gcdignlgbr2a-big-TOC.txt])
E           Failed to autodetect and load schema, try again with --logger-log-level scylla-sstable=debug to learn more or provide the schema source manually

../scylla/.local/lib/python3.11/site-packages/ccmlib/scylla_node.py:1511: ToolError
```

[logs-full.release.005.zip](https://github.com/scylladb/scylladb/files/15398814/logs-full.release.005.zip)

**Comment by denesb** [comment]
scylla-sstable raced with compaction.

**Comment by denesb** [comment]
https://github.com/scylladb/scylladb/issues/17869 will fix this.

**Comment by Annamikhlin** [comment]
seen on https://jenkins.scylladb.com/job/scylla-6.0/job/next/139/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Build___x86___dtest_with_topology_changes___test_rebuild_node_streaming/

Can be backported to 6.0?

**Comment by denesb** [comment]
> seen on https://jenkins.scylladb.com/job/scylla-6.0/job/next/139/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Build___x86___dtest_with_topology_changes___test_rebuild_node_streaming/
> 
> Can be backported to 6.0?

Hmm, the fix (https://github.com/scylladb/scylladb/issues/17869) is not something we normally backport. It is a new way to obtain the schema, so it is a new feature, rather than a bugfix. It is also not a regression -- the tool didn't have a better way in the past to read the schema.
So far we have only backported scylla-sstable related changes to allow replacing broken java tools, and to add tooling support for EAR.
That said, if this becomes a big pain in 6.0, we *can* consider this for backport.

**Comment by mykaul** [comment]
> > seen on https://jenkins.scylladb.com/job/scylla-6.0/job/next/139/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/Build___x86___dtest_with_topology_changes___test_rebuild_node_streaming/
> > Can be backported to 6.0?
> 
> Hmm, the fix (#17869) is not something we normally backport. It is a new way to obtain the schema, so it is a new feature, rather than a bugfix. It is also not a regression -- the tool didn't have a better way in the past to read the schema. So far we have only backported scylla-sstable related changes to allow replacing broken java tools, and to add tooling support for EAR. That said, if this becomes a big pain in 6.0, we _can_ consider this for backport.

I think we should, if it's safe. It'll help us in CI.

**Comment by denesb** [comment]
> I think we should, if it's safe. It'll help us in CI.

Ok, I'll backport it.
